full/src/main/java/apoc/load/LoadDirectory.java

```diff
@@ -2,35 +2,86 @@
 
 import apoc.Extended;
 import apoc.result.StringResult;
-import apoc.util.FileUtils;
 import apoc.util.Util;
 import org.apache.commons.io.filefilter.TrueFileFilter;
 import org.apache.commons.io.filefilter.WildcardFileFilter;
+import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.Transaction;
 import org.neo4j.logging.Log;
+import org.neo4j.procedure.Mode;
 import org.neo4j.procedure.Procedure;
 import org.neo4j.procedure.Context;
 import org.neo4j.procedure.Description;
 import org.neo4j.procedure.Name;
 
 import java.io.File;
 import java.io.IOException;
-import java.net.URI;
-import java.nio.file.Paths;
 import java.util.Map;
 import java.util.Collection;
 import java.util.stream.Stream;
 
 import static apoc.ApocConfig.apocConfig;
-import static apoc.util.FileUtils.isImportUsingNeo4jConfig;
-import static org.eclipse.jetty.util.URIUtil.encodePath;
-import static org.apache.commons.lang3.StringUtils.replaceOnce;
+import static apoc.util.FileUtils.checkIfUrlEmptyAndGetFileUrl;
+import static apoc.util.FileUtils.getPathDependingOnUseNeo4jConfig;
+import static apoc.util.FileUtils.getPathFromUrlString;
 
 @Extended
 public class LoadDirectory {
 
     @Context
     public Log log;
 
+    @Context
+    public GraphDatabaseService db;
+
+    @Context
+    public LoadDirectoryHandler loadDirectoryHandler;
+
+    @Context
+    public Transaction tx;
+
+
+    @Procedure(name="apoc.load.directory.async.add", mode = Mode.WRITE)
+    @Description("apoc.load.directory.async.add(name, cypher, pattern, urlDir, {}) YIELD name, status, pattern, cypher, urlDir, config, error - Add or replace a folder listener with a specific name, pattern and url directory that execute the specified cypher query when an event is triggered and return listener list")
+    public Stream<LoadDirectoryItem.LoadDirectoryResult> add(@Name("name") String name,
+                                                             @Name("cypher") String cypher,
+                                                             @Name(value = "pattern", defaultValue = "*") String pattern,
+                                                             @Name(value = "urlDir", defaultValue = "") String urlDir,
+                                                             @Name(value = "config", defaultValue = "{}") Map<String, Object> config) {
+        apocConfig().checkReadAllowed();
```

Comment:
Per favore dopo questo metodo aggiungi una validazione della query come fatto in `Periodic.java`, devi validare anche il fatto che la query sia di `WRITE` o `READ_WRITE`